sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/GpuWindowInPandasExecBase.scala

```diff
@@ -78,96 +77,265 @@ abstract class GpuWindowInPandasExecMetaBase(
+      (sBatch.getColumnarBatch(), lens)
     } else {
       val batch = wrapped.next()
       inputBatches += 1
       inputRows += batch.numRows()
-      if (batch.numRows() > 0 && batch.numCols() > 0 && partitionSpec.nonEmpty) {
-        val partitionIndices = partitionSpec.indices
-        // 1) Calculate the split indices via cudf.Table.groupBy and aggregation Count.
-        //   a) Compute the count number for each group in a batch, including null values.
-        //   b) Restore the original order (Ascending with NullsFirst) defined in the plan,
-        //      since cudf.Table.groupBy will break the original order.
-        //   c) The 'count' column can be used to calculate the indices for split.
-        val cntTable = withResource(GpuProjectExec.project(batch, partitionSpec)) { projected =>
-          withResource(GpuColumnVector.from(projected)) { table =>
-            table
-              .groupBy(partitionIndices:_*)
-              .aggregate(Aggregation.count(true).onColumn(0))
-          }
+      val groupLens = genGroupInfoFromBatch(batch)
+      if (isSplitGroups && groupLens.size > 1) {
+        // Split the groups when being required and there are more than one group.
+        val groupBatchesAndSizes = splitBatchPerGroups(batch, groupLens)
+          .zip(groupLens.map(Seq(_)))
+        val tails = groupBatchesAndSizes.tail.map {
+          case (sb, gpInfo) =>
+            (SpillableColumnarBatch(sb, SpillPriorities.ACTIVE_ON_DECK_PRIORITY, spillCallback),
+              gpInfo)
         }
-        val orderedTable = withResource(cntTable) { table =>
-          table.orderBy(partitionIndices.map(id => Table.asc(id, true)): _*)
+        // Return the first one and enqueue others
+        queueBatches.enqueue(tails: _*)
+        groupBatchesAndSizes.head
+      } else {
+        (batch, groupLens)
+      }
+    }
+    (nextBatch, buildInfoBatchForPython(nextBatch, gpLens))
+  }
+
+  private[this] def splitBatchPerGroups(
+      batch: ColumnarBatch,
+      gpLens: Seq[Int]): Seq[ColumnarBatch] = {
+    val splitIndices = gpLens.dropRight(1).scan(0)(_+_).tail
+    // Close the original big batch
+    withResource(batch) { cb =>
+      withResource(GpuColumnVector.from(cb)) { table =>
+        withResource(table.contiguousSplit(splitIndices: _*)) { tables =>
+          val dataTypes = GpuColumnVector.extractTypes(batch)
+          tables.safeMap(t => GpuColumnVectorFromBuffer.from(t, dataTypes))
         }
-        val (countHostCol, numRows) = withResource(orderedTable) { table =>
-          // Yes copying the data to host, it would be OK since just copying the aggregated
-          // count column.
-          (table.getColumn(table.getNumberOfColumns - 1).copyToHost(), table.getRowCount)
+      }
+    }
+  }
+
+  private[this] def buildInfoBatchForPython(
+      batch: ColumnarBatch,
+      gpLens: Seq[_]): ColumnarBatch = {
+
+    // The rolling size may need tuning.
+    val rollingSize = 20000
```

Comment:
Where exactly did the 20,000 come from? Also why are we using the number of groups? Why is this not related to the amount of data being sent or number of rows being sent instead of groups. It feels like it would be close to impossible to predict how much memory the python side would need to do the processing.